File: code/classes/pattern-matching.scala
object PatternMatching extends App {
  def describe(obj: Any): String = obj match {
    case 1 =&gt; "One"
    case "Hello" =&gt; "Greeting"
    case _: Long =&gt; "Long number"
    case _: Any if !obj.isInstanceOf[String] =&gt; "Not a string"
    case _ =&gt; "Unknown"
  }
  
  println(describe(1))
  println(describe("Hello"))
  println(describe(2L))
}

File: code/classes/pattern-matching.java
// Pattern matching in Java (using instanceof and switch expressions - Java 14+)
public class PatternMatching {
    public static String describe(Object obj) {
        if (obj instanceof Integer i && i == 1) {
            return "One";
        } else if (obj instanceof String s && s.equals("Hello")) {
            return "Greeting";
        } else if (obj instanceof Long) {
            return "Long number";
        } else {
            return "Unknown";
        }
    }
    
    public static void main(String[] args) {
        System.out.println(describe(1));
        System.out.println(describe("Hello"));
        System.out.println(describe(2L));
    }
}